File: Android/app/src/main/java/com/azhegezhege/zhuzhiliao/physics/ToySimulation.kt
package com.azhegezhege.zhuzhiliao.physics

import com.azhegezhege.zhuzhiliao.math.Vec3
import com.azhegezhege.zhuzhiliao.motion.ShakeDrive
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin

data class MotionInput(
    val anchorAcceleration: Vec3,
    val gravityDirection: Vec3,
    val rotationRate: Vec3,
    val anchorTarget: Vec3? = null,
    val shakeDrive: ShakeDrive = ShakeDrive.INACTIVE,
) {
    companion object {
        val ZERO = MotionInput(Vec3.ZERO, Vec3.ZERO, Vec3.ZERO)
    }
}

data class ToyPhysicsState(
    var position: Vec3,
    var velocity: Vec3,
    var ropeLength: Float,
    var angularVelocity: Vec3,
    var tension: Float,
    var activity: Float,
    var orbitCoherence: Float = 0f,
    var anchorOffset: Vec3 = Vec3.ZERO,
)

data class ToyPhysicsConfiguration(
    val ropeLength: Float = 1.65f,
    val fixedTimeStep: Float = 1f / 240f,
    val gravityMagnitude: Float = 7.67f,
    val motionAccelerationScale: Float = 21f,
    val ropeStiffness: Float = 2_600f,
    val ropeDamping: Float = 14f,
    val airDrag: Float = 0.35f,
    val maximumStretchRatio: Float = Float.MAX_VALUE,
    val soundThresholdRPS: Float = 1.1f,
    val soundRampRPS: Float = 2.6f,
    val orbitQualityThreshold: Float = 0.55f,
    val minimumShakeRPS: Float = 1.2f,
    val maximumShakeRPS: Float = 3.4f,
    val shakeDriveResponseTime: Float = 0.35f,
    val maximumShakeAcceleration: Float = 35f,
    val coherenceThreshold: Float = 0.60f,
)

data class SimulationFrame(
    val state: ToyPhysicsState,
    val revolutionsPerSecond: Float,
    val phase: Float,
    val completedWahs: Int,
)

class ToySimulation(
    private val configuration: ToyPhysicsConfiguration = ToyPhysicsConfiguration(),
    initialState: ToyPhysicsState? = null,
) {
    private var accumulatedTime = 0.0
    private val orbitCounter = OrbitCounter()
    private var phase = 0f
    private var stableOrbitAxis = Vec3(0f, 0f, 1f)
    private var hasStableOrbitAxis = false
    private var lastDriveAxis: Vec3? = null

    var state = initialState ?: ToyPhysicsState(
        position = Vec3(0.06f, -configuration.ropeLength * 0.92f, 0f),
        velocity = Vec3(0.22f, 0f, 0f),
        ropeLength = configuration.ropeLength,
        angularVelocity = Vec3.ZERO,
        tension = 0f,
        activity = 0f,
    )
        private set

    fun reset(gravityDirection: Vec3 = Vec3(0f, -1f, 0f)) {
        val direction = gravityDirection.normalized(Vec3(0f, -1f, 0f))
        state = ToyPhysicsState(
            position = direction * configuration.ropeLength * 0.92f,
            velocity = Vec3(0.22f, 0f, 0f),
            ropeLength = configuration.ropeLength,
            angularVelocity = Vec3.ZERO,
            tension = 0f,
            activity = 0f,
        )
        accumulatedTime = 0.0
        orbitCounter.reset()
        phase = 0f
        stableOrbitAxis = Vec3(0f, 0f, 1f)
        hasStableOrbitAxis = false
        lastDriveAxis = null
    }

    fun advance(input: MotionInput, deltaTime: Double): SimulationFrame {
        accumulatedTime += deltaTime.coerceIn(0.0, 0.05)
        val step = configuration.fixedTimeStep.toDouble()
        var completedWahs = 0
        while (accumulatedTime + 1e-8 >= step) {
            completedWahs += integrate(input, configuration.fixedTimeStep)
            accumulatedTime -= step
        }
        return SimulationFrame(
            state = state.copy(),
            revolutionsPerSecond = state.angularVelocity.length / FULL_TURN,
            phase = phase,
            completedWahs = completedWahs,
        )
    }

    private fun integrate(input: MotionInput, timeStep: Float): Int {
        var anchorVelocity = Vec3.ZERO
        input.anchorTarget?.let { target ->
            val previousAnchor = state.anchorOffset
            val response = 1f - exp(-timeStep * 26f)
            state.anchorOffset += (target - state.anchorOffset) * response
            val anchorDelta = state.anchorOffset - previousAnchor
            state.position -= anchorDelta
            anchorVelocity = anchorDelta / timeStep
        }

        synchronizeDriveDirection(input.shakeDrive)
        val gravity = input.gravityDirection.normalized(Vec3.ZERO) * configuration.gravityMagnitude
        var acceleration = gravity - input.anchorAcceleration * configuration.motionAccelerationScale -
            state.velocity * configuration.airDrag
        acceleration += shakeAssistance(input.shakeDrive, anchorVelocity)

        val distance = state.position.length
        if (distance > configuration.ropeLength && distance > Vec3.EPSILON) {
            val direction = state.position / distance
            val radialVelocity = state.velocity.dot(direction)
            val pull = max(
                0f,
                configuration.ropeStiffness * (distance - configuration.ropeLength) +
                    configuration.ropeDamping * radialVelocity,
            )
            acceleration -= direction * pull
        }

        state.velocity += acceleration * timeStep
        state.position += state.velocity * timeStep

        val maximumDistance = configuration.ropeLength * configuration.maximumStretchRatio
        val integratedDistance = state.position.length
        if (integratedDistance > maximumDistance && integratedDistance > Vec3.EPSILON) {
            val direction = state.position / integratedDistance
            state.position = direction * maximumDistance
            val outwardSpeed = state.velocity.dot(direction)
            if (outwardSpeed > 0f) state.velocity -= direction * outwardSpeed
        }

        val orbitQuality = updateDerivedState(
            anchorVelocity,
            input.shakeDrive.orbitAxis.takeIf { input.shakeDrive.isActive },
            timeStep,
        )
        return updateRevolutionCount(orbitQuality, timeStep)
    }

    private fun synchronizeDriveDirection(drive: ShakeDrive) {
        if (!drive.isActive) return
        val axis = drive.orbitAxis.normalized(ShakeDrive.DEFAULT_ORBIT_AXIS)
        val previous = lastDriveAxis
        if (previous != null && previous.dot(axis) < 0.5f) {
            orbitCounter.reset()
            state.orbitCoherence = 0f
            state.activity = 0f
            stableOrbitAxis = axis
            hasStableOrbitAxis = true
        } else if (previous == null && state.angularVelocity.length < 0.5f) {
            stableOrbitAxis = axis
            hasStableOrbitAxis = true
        }
        lastDriveAxis = axis
    }

    private fun shakeAssistance(drive: ShakeDrive, anchorVelocity: Vec3): Vec3 {
        if (!drive.isActive) return Vec3.ZERO
        val axis = drive.orbitAxis.normalized(ShakeDrive.DEFAULT_ORBIT_AXIS)
        val radiusDirection = state.position.normalized(Vec3(0f, -1f, 0f))
        var tangent = axis.cross(radiusDirection)
        if (tangent.lengthSquared < Vec3.EPSILON) {
            tangent = axis.cross(if (abs(axis.y) < 0.9f) Vec3.UP else Vec3(1f, 0f, 0f))
        }
        tangent = tangent.normalized(Vec3(1f, 0f, 0f))

        val intensity = drive.intensity.coerceIn(0f, 1f)
        val targetRps = configuration.minimumShakeRPS +
            (configuration.maximumShakeRPS - configuration.minimumShakeRPS) * intensity
        val effectiveRadius = state.position.length.coerceIn(
            configuration.ropeLength * 0.55f,
            configuration.ropeLength,
        )
        val targetTangentialSpeed = targetRps * FULL_TURN * effectiveRadius
        val relativeVelocity = state.velocity - anchorVelocity
        val currentTangentialSpeed = relativeVelocity.dot(tangent)
        val speedError = max(0f, targetTangentialSpeed - currentTangentialSpeed)
        val requestedAcceleration = if (configuration.shakeDriveResponseTime > 0f) {
            speedError / configuration.shakeDriveResponseTime
        } else {
            configuration.maximumShakeAcceleration
        }
        val accelerationLimit = configuration.maximumShakeAcceleration * (0.35f + 0.65f * intensity)
        return tangent * min(requestedAcceleration, accelerationLimit)
    }

    private fun updateDerivedState(
        anchorVelocity: Vec3,
        preferredDriveAxis: Vec3?,
        timeStep: Float,
    ): Float {
        val distanceSquared = state.position.lengthSquared
        val relativeVelocity = state.velocity - anchorVelocity
        val measuredAngularVelocity = if (distanceSquared > Vec3.EPSILON) {
            state.position.cross(relativeVelocity) / distanceSquared
        } else Vec3.ZERO

        state.angularVelocity += (measuredAngularVelocity - state.angularVelocity) * min(1f, timeStep * 9f)
        val measuredAngularSpeed = measuredAngularVelocity.length
        var targetCoherence = 0f
        if (measuredAngularSpeed > 0.08f) {
            val measuredAxis = measuredAngularVelocity / measuredAngularSpeed
            if (!hasStableOrbitAxis) {
                stableOrbitAxis = measuredAxis
                hasStableOrbitAxis = true
            } else {
                val referenceAxis = preferredDriveAxis?.normalized(stableOrbitAxis) ?: stableOrbitAxis
                val alignment = measuredAxis.dot(referenceAxis)
                if (alignment >= 0.25f) {
                    targetCoherence = smoothStep(0.65f, 0.98f, alignment)
                    stableOrbitAxis = if (preferredDriveAxis != null) {
                        referenceAxis
                    } else {
                        (stableOrbitAxis * 0.88f + measuredAxis * 0.12f).normalized(measuredAxis)
                    }
                } else {
                    state.orbitCoherence = 0f
                    orbitCounter.reset()
                    if (alignment < -0.25f || preferredDriveAxis == null) stableOrbitAxis = measuredAxis
                }
            }
        }

        val coherenceResponse = if (targetCoherence > state.orbitCoherence) 4.5f else 12f
        state.orbitCoherence += (targetCoherence - state.orbitCoherence) * min(1f, timeStep * coherenceResponse)

        val angularSpeed = state.angularVelocity.length
        if (measuredAngularSpeed <= Vec3.EPSILON && angularSpeed > Vec3.EPSILON) {
            state.angularVelocity = stableOrbitAxis * angularSpeed
        }

        val distance = state.position.length
        state.tension = ((distance / configuration.ropeLength - 0.88f) / 0.12f).coerceIn(0f, 1f)
        val speed = relativeVelocity.length
        val crossMagnitude = state.position.cross(relativeVelocity).length
        val orbitQuality = if (distance > Vec3.EPSILON && speed > Vec3.EPSILON) {
            crossMagnitude / (distance * speed)
        } else 0f

        val rps = angularSpeed / FULL_TURN
        val drive = ((rps - configuration.soundThresholdRPS) / configuration.soundRampRPS).coerceIn(0f, 1f)
        val qualityGate = smoothStep(configuration.orbitQualityThreshold, 0.92f, orbitQuality)
        val coherenceGate = smoothStep(configuration.coherenceThreshold, 0.88f, state.orbitCoherence)
        val targetActivity = drive.pow(1.25f) * state.tension * qualityGate * coherenceGate
        val response = if (targetActivity > state.activity) 10f else 3.2f
        state.activity += (targetActivity - state.activity) * min(1f, timeStep * response)
        phase += angularSpeed * timeStep * dominantAxisSign(state.angularVelocity)
        phase %= FULL_TURN
        return orbitQuality
    }

    private fun updateRevolutionCount(orbitQuality: Float, timeStep: Float): Int {
        val angularSpeed = state.angularVelocity.length
        return orbitCounter.update(
            angleDelta = angularSpeed * timeStep,
            axis = if (angularSpeed > Vec3.EPSILON) state.angularVelocity / angularSpeed else stableOrbitAxis,
            isQualified = state.activity > 0.3f && state.tension > 0.8f &&
                state.orbitCoherence >= configuration.coherenceThreshold &&
                orbitQuality > configuration.orbitQualityThreshold,
        )
    }

    companion object {
        val FULL_TURN = (2.0 * PI).toFloat()
    }
}

object WebRopeShape {
    fun points(anchor: Vec3, bob: Vec3, ropeLength: Float, count: Int = 25): List<Vec3> {
        require(count >= 2)
        val slack = max(0f, ropeLength - anchor.distance(bob))
        val sag = slack * 0.55f
        return List(count) { index ->
            when (index) {
                0 -> anchor
                count - 1 -> bob
                else -> {
                    val fraction = index.toFloat() / (count - 1).toFloat()
                    val point = anchor + (bob - anchor) * fraction
                    point.copy(y = point.y - sag * sin(PI.toFloat() * fraction) * (0.75f + 0.25f * fraction))
                }
            }
        }
    }
}

private fun dominantAxisSign(value: Vec3): Float {
    val dominant = when {
        abs(value.x) >= abs(value.y) && abs(value.x) >= abs(value.z) -> value.x
        abs(value.y) >= abs(value.z) -> value.y
        else -> value.z
    }
    return if (dominant < 0f) -1f else 1f
}

private fun smoothStep(edge0: Float, edge1: Float, value: Float): Float {
    if (edge1 <= edge0) return if (value >= edge1) 1f else 0f
    val fraction = ((value - edge0) / (edge1 - edge0)).coerceIn(0f, 1f)
    return fraction * fraction * (3f - 2f * fraction)
}

File: iOS/ZhuZhiLiao/Physics/ToySimulation.swift
import Foundation
import simd

struct MotionInput: Equatable, Sendable {
    var anchorAcceleration: SIMD3<Float>
    var gravityDirection: SIMD3<Float>
    var rotationRate: SIMD3<Float>
    /// 网页版的鼠标/触摸输入：杆梢相对初始位置的目标位移。
    /// `nil` 表示杆梢保持当前位置，仅使用设备惯性输入。
    var anchorTarget: SIMD3<Float>? = nil
    /// 任意方向持续摇动产生的稳定切向助力；触摸模式始终传入 `.inactive`。
    var shakeDrive: ShakeDrive = .inactive

    static let zero = MotionInput(
        anchorAcceleration: .zero,
        gravityDirection: .zero,
        rotationRate: .zero
    )
}

struct ToyPhysicsState: Equatable, Sendable {
    /// 竹筒相对杆梢的位置。竹筒速度仍位于世界坐标中，与网页版一致。
    var position: SIMD3<Float>
    var velocity: SIMD3<Float>
    var ropeLength: Float
    var angularVelocity: SIMD3<Float>
    var tension: Float
    var activity: Float
    /// 角速度方向在一段时间内保持一致的程度，防止往复摆动误发声。
    var orbitCoherence: Float = 0
    /// 杆梢相对初始场景锚点的位置。
    var anchorOffset: SIMD3<Float> = .zero
}

struct ToyPhysicsConfiguration: Equatable, Sendable {
    var ropeLength: Float
    var fixedTimeStep: Float
    var gravityMagnitude: Float
    var motionAccelerationScale: Float
    var ropeStiffness: Float
    var ropeDamping: Float
    var airDrag: Float
    var maximumStretchRatio: Float
    var soundThresholdRPS: Float
    var soundRampRPS: Float
    var orbitQualityThreshold: Float
    var minimumShakeRPS: Float
    var maximumShakeRPS: Float
    var shakeDriveResponseTime: Float
    var maximumShakeAcceleration: Float
    var coherenceThreshold: Float

    static let live = ToyPhysicsConfiguration(
        ropeLength: 1.65,
        fixedTimeStep: 1 / 240,
        gravityMagnitude: 7.67,
        motionAccelerationScale: 21,
        ropeStiffness: 2_600,
        ropeDamping: 14,
        airDrag: 0.35,
        // 网页版不硬截断绳长，过伸完全由弹簧力拉回；保留字段只供
        // 测试或特殊配置设置数值安全上限。
        maximumStretchRatio: .greatestFiniteMagnitude,
        soundThresholdRPS: 1.1,
        soundRampRPS: 2.6,
        orbitQualityThreshold: 0.55,
        minimumShakeRPS: 1.2,
        maximumShakeRPS: 3.4,
        shakeDriveResponseTime: 0.35,
        maximumShakeAcceleration: 35,
        coherenceThreshold: 0.60
    )

    static let testing = live

    static let inertialTesting = ToyPhysicsConfiguration(
        ropeLength: 1,
        fixedTimeStep: 1 / 240,
        gravityMagnitude: 0,
        motionAccelerationScale: 1,
        ropeStiffness: 0,
        ropeDamping: 0,
        airDrag: 0,
        maximumStretchRatio: 10,
        soundThresholdRPS: 1.1,
        soundRampRPS: 2.6,
        orbitQualityThreshold: 0.65,
        minimumShakeRPS: 1.2,
        maximumShakeRPS: 3.4,
        shakeDriveResponseTime: 0.35,
        maximumShakeAcceleration: 35,
        coherenceThreshold: 0.60
    )
}

struct SimulationFrame: Equatable, Sendable {
    var state: ToyPhysicsState
    var revolutionsPerSecond: Float
    var phase: Float
    var completedWahs: Int
}

struct ToySimulation: Sendable {
    private let configuration: ToyPhysicsConfiguration
    private var accumulatedTime: Double = 0
    private var orbitCounter = OrbitCounter()
    private var phase: Float = 0
    private var stableOrbitAxis = SIMD3<Float>(0, 0, 1)
    private var hasStableOrbitAxis = false
    private var lastDriveAxis: SIMD3<Float>?

    private(set) var state: ToyPhysicsState

    init(configuration: ToyPhysicsConfiguration = .live) {
        self.configuration = configuration
        state = ToyPhysicsState(
            position: SIMD3<Float>(0.06, -configuration.ropeLength * 0.92, 0),
            velocity: SIMD3<Float>(0.22, 0, 0),
            ropeLength: configuration.ropeLength,
            angularVelocity: .zero,
            tension: 0,
            activity: 0
        )
    }

    init(state: ToyPhysicsState, configuration: ToyPhysicsConfiguration) {
        self.state = state
        self.configuration = configuration
    }

    mutating func reset(gravityDirection: SIMD3<Float> = SIMD3<Float>(0, -1, 0)) {
        let direction = gravityDirection.normalized(or: SIMD3<Float>(0, -1, 0))
        state = ToyPhysicsState(
            position: direction * configuration.ropeLength * 0.92,
            velocity: SIMD3<Float>(0.22, 0, 0),
            ropeLength: configuration.ropeLength,
            angularVelocity: .zero,
            tension: 0,
            activity: 0
        )
        accumulatedTime = 0
        orbitCounter.reset()
        phase = 0
        stableOrbitAxis = SIMD3<Float>(0, 0, 1)
        hasStableOrbitAxis = false
        lastDriveAxis = nil
    }

    mutating func advance(input: MotionInput, deltaTime: TimeInterval) -> SimulationFrame {
        accumulatedTime += min(max(deltaTime, 0), 0.05)
        let step = Double(configuration.fixedTimeStep)
        var completedWahs = 0

        while accumulatedTime + 1e-8 >= step {
            completedWahs += integrate(input: input, timeStep: configuration.fixedTimeStep)
            accumulatedTime -= step
        }

        return SimulationFrame(
            state: state,
            revolutionsPerSecond: simd_length(state.angularVelocity) / (2 * .pi),
            phase: phase,
            completedWahs: completedWahs
        )
    }

    private mutating func integrate(input: MotionInput, timeStep: Float) -> Int {
        // 网页版不是把竹筒直接拖到指针下，而是让杆梢以 26/s 的一阶响应
        // 追随目标。先移动锚点，再保持竹筒的世界位置不变，就会自然产生
        // 绳子滞后、收紧和回摆。
        var anchorVelocity = SIMD3<Float>.zero
        if let target = input.anchorTarget {
            let previousAnchor = state.anchorOffset
            let response = 1 - exp(-timeStep * 26)
            state.anchorOffset += (target - state.anchorOffset) * response
            let anchorDelta = state.anchorOffset - previousAnchor
            state.position -= anchorDelta
            anchorVelocity = anchorDelta / timeStep
        }

        synchronizeDriveDirection(input.shakeDrive)

        let gravity = input.gravityDirection.normalized(or: .zero) * configuration.gravityMagnitude
        var acceleration = gravity
            - input.anchorAcceleration * configuration.motionAccelerationScale
            - state.velocity * configuration.airDrag
        acceleration += shakeAssistance(
            input.shakeDrive,
            anchorVelocity: anchorVelocity
        )

        let distance = simd_length(state.position)
        if distance > configuration.ropeLength, distance > 0.000_001 {
            let direction = state.position / distance
            let radialVelocity = simd_dot(state.velocity, direction)
            let pull = max(
                0,
                configuration.ropeStiffness * (distance - configuration.ropeLength)
                    + configuration.ropeDamping * radialVelocity
            )
            acceleration -= direction * pull
        }

        state.velocity += acceleration * timeStep
        state.position += state.velocity * timeStep

        let maximumDistance = configuration.ropeLength * configuration.maximumStretchRatio
        let integratedDistance = simd_length(state.position)
        if integratedDistance > maximumDistance, integratedDistance > 0.000_001 {
            let direction = state.position / integratedDistance
            state.position = direction * maximumDistance
            let outwardSpeed = simd_dot(state.velocity, direction)
            if outwardSpeed > 0 {
                state.velocity -= direction * outwardSpeed
            }
        }

        let orbitQuality = updateDerivedState(
            anchorVelocity: anchorVelocity,
            preferredDriveAxis: input.shakeDrive.isActive
                ? input.shakeDrive.orbitAxis
                : nil,
            timeStep: timeStep
        )
        return updateRevolutionCount(
            orbitQuality: orbitQuality,
            timeStep: timeStep
        )
    }

    private mutating func synchronizeDriveDirection(_ drive: ShakeDrive) {
        guard drive.isActive else { return }
        let axis = drive.orbitAxis.normalized(or: ShakeDrive.defaultOrbitAxis)

        if let lastDriveAxis, simd_dot(lastDriveAxis, axis) < 0.5 {
            orbitCounter.reset()
            state.orbitCoherence = 0
            state.activity = 0
            stableOrbitAxis = axis
            hasStableOrbitAxis = true
        } else if lastDriveAxis == nil,
                  simd_length(state.angularVelocity) < 0.5 {
            stableOrbitAxis = axis
            hasStableOrbitAxis = true
        }
        lastDriveAxis = axis
    }

    private func shakeAssistance(
        _ drive: ShakeDrive,
        anchorVelocity: SIMD3<Float>
    ) -> SIMD3<Float> {
        guard drive.isActive else { return .zero }

        let axis = drive.orbitAxis.normalized(or: ShakeDrive.defaultOrbitAxis)
        let radiusDirection = state.position.normalized(or: SIMD3<Float>(0, -1, 0))
        var tangent = simd_cross(axis, radiusDirection)
        if simd_length_squared(tangent) < 0.000_001 {
            let fallback = abs(axis.y) < 0.9
                ? SIMD3<Float>(0, 1, 0)
                : SIMD3<Float>(1, 0, 0)
            tangent = simd_cross(axis, fallback)
        }
        tangent = tangent.normalized(or: SIMD3<Float>(1, 0, 0))

        let intensity = clamp(drive.intensity, 0, 1)
        let targetRPS = configuration.minimumShakeRPS
            + (configuration.maximumShakeRPS - configuration.minimumShakeRPS) * intensity
        let effectiveRadius = clamp(
            simd_length(state.position),
            configuration.ropeLength * 0.55,
            configuration.ropeLength
        )
        let targetTangentialSpeed = targetRPS * 2 * .pi * effectiveRadius
        let relativeVelocity = state.velocity - anchorVelocity
        let currentTangentialSpeed = simd_dot(relativeVelocity, tangent)
        let speedError = max(0, targetTangentialSpeed - currentTangentialSpeed)
        let requestedAcceleration = configuration.shakeDriveResponseTime > 0
            ? speedError / configuration.shakeDriveResponseTime
            : configuration.maximumShakeAcceleration
        let accelerationLimit = configuration.maximumShakeAcceleration
            * (0.35 + 0.65 * intensity)
        return tangent * min(requestedAcceleration, accelerationLimit)
    }

    @discardableResult
    private mutating func updateDerivedState(
        anchorVelocity: SIMD3<Float>,
        preferredDriveAxis: SIMD3<Float>?,
        timeStep: Float
    ) -> Float {
        let distanceSquared = simd_length_squared(state.position)
        let measuredAngularVelocity: SIMD3<Float>
        let relativeVelocity = state.velocity - anchorVelocity
        if distanceSquared > 0.000_001 {
            measuredAngularVelocity = simd_cross(state.position, relativeVelocity) / distanceSquared
        } else {
            measuredAngularVelocity = .zero
        }

        // 网页版用 9/s 的一阶滤波稳定绳方向角速度；向量形式让三维轨道
        // 也保留同样的起转、换向和松手余摆手感。
        state.angularVelocity += (measuredAngularVelocity - state.angularVelocity)
            * min(1, timeStep * 9)

        let measuredAngularSpeed = simd_length(measuredAngularVelocity)
        var targetCoherence: Float = 0
        if measuredAngularSpeed > 0.08 {
            let measuredAxis = measuredAngularVelocity / measuredAngularSpeed
            if !hasStableOrbitAxis {
                stableOrbitAxis = measuredAxis
                hasStableOrbitAxis = true
            } else {
                let referenceAxis = preferredDriveAxis?.normalized(or: stableOrbitAxis)
                    ?? stableOrbitAxis
                let alignment = simd_dot(measuredAxis, referenceAxis)
                if alignment >= 0.25 {
                    targetCoherence = smoothStep(
                        edge0: 0.65,
                        edge1: 0.98,
                        value: alignment
                    )
                    if preferredDriveAxis != nil {
                        stableOrbitAxis = referenceAxis
                    } else {
                        stableOrbitAxis = (stableOrbitAxis * 0.88 + measuredAxis * 0.12)
                            .normalized(or: measuredAxis)
                    }
                } else {
                    state.orbitCoherence = 0
                    orbitCounter.reset()
                    if alignment < -0.25 || preferredDriveAxis == nil {
                        stableOrbitAxis = measuredAxis
                    }
                }
            }
        }

        let coherenceResponse: Float = targetCoherence > state.orbitCoherence ? 4.5 : 12
        state.orbitCoherence += (targetCoherence - state.orbitCoherence)
            * min(1, timeStep * coherenceResponse)

        let angularSpeed = simd_length(state.angularVelocity)
        if measuredAngularSpeed <= 0.000_001, angularSpeed > 0.000_001 {
            state.angularVelocity = stableOrbitAxis * angularSpeed
        }

        let distance = simd_length(state.position)
        state.tension = clamp((distance / configuration.ropeLength - 0.88) / 0.12, 0, 1)
        let speed = simd_length(relativeVelocity)
        let crossMagnitude = simd_length(simd_cross(state.position, relativeVelocity))
        let orbitQuality = distance > 0.000_001 && speed > 0.000_001
            ? crossMagnitude / (distance * speed)
            : 0

        let rps = angularSpeed / (2 * .pi)
        let drive = clamp(
            (rps - configuration.soundThresholdRPS) / configuration.soundRampRPS,
            0,
            1
        )
        let qualityGate = smoothStep(
            edge0: configuration.orbitQualityThreshold,
            edge1: 0.92,
            value: orbitQuality
        )
        let coherenceGate = smoothStep(
            edge0: configuration.coherenceThreshold,
            edge1: 0.88,
            value: state.orbitCoherence
        )
        let targetActivity = pow(drive, 1.25)
            * state.tension
            * qualityGate
            * coherenceGate
        let response: Float = targetActivity > state.activity ? 10 : 3.2
        state.activity += (targetActivity - state.activity) * min(1, timeStep * response)
        phase += angularSpeed * timeStep * dominantAxisSign(state.angularVelocity)
        phase.formTruncatingRemainder(dividingBy: 2 * .pi)
        return orbitQuality
    }

    private mutating func updateRevolutionCount(
        orbitQuality: Float,
        timeStep: Float
    ) -> Int {
        let angularSpeed = simd_length(state.angularVelocity)
        return orbitCounter.update(
            angleDelta: angularSpeed * timeStep,
            axis: angularSpeed > 0.000_001 ? state.angularVelocity / angularSpeed : stableOrbitAxis,
            isQualified: state.activity > 0.3
                && state.tension > 0.8
                && state.orbitCoherence >= configuration.coherenceThreshold
                && orbitQuality > configuration.orbitQualityThreshold
        )
    }
}

/// 与网页 3D 层 `catenaryPoints` 相同的绳形。物理端点来自定步长模拟，
/// 绳中段只消费松弛量，因此收紧时立刻拉直，松绳时自然向下垂落。
enum WebRopeShape {
    static func points(
        anchor: SIMD3<Float>,
        bob: SIMD3<Float>,
        ropeLength: Float,
        count: Int = 25
    ) -> [SIMD3<Float>] {
        var result: [SIMD3<Float>] = []
        update(
            &result,
            anchor: anchor,
            bob: bob,
            ropeLength: ropeLength,
            count: count
        )
        return result
    }

    static func update(
        _ points: inout [SIMD3<Float>],
        anchor: SIMD3<Float>,
        bob: SIMD3<Float>,
        ropeLength: Float,
        count: Int = 25
    ) {
        precondition(count >= 2)
        let slack = max(0, ropeLength - simd_distance(anchor, bob))
        let sag = slack * 0.55
        points.removeAll(keepingCapacity: true)
        points.reserveCapacity(count)

        for index in 0..<count {
            if index == 0 {
                points.append(anchor)
                continue
            }
            if index == count - 1 {
                points.append(bob)
                continue
            }
            let fraction = Float(index) / Float(count - 1)
            var point = simd_mix(
                anchor,
                bob,
                SIMD3<Float>(repeating: fraction)
            )
            point.y -= sag
                * sin(.pi * fraction)
                // 网页函数从竹筒端采样到杆梢端；这里数组方向相反，
                // 因此把原式 (1 - .25*t) 反向后使用。
                * (0.75 + 0.25 * fraction)
            points.append(point)
        }
    }
}

private func clamp(_ value: Float, _ lower: Float, _ upper: Float) -> Float {
    min(max(value, lower), upper)
}

private func dominantAxisSign(_ value: SIMD3<Float>) -> Float {
    let magnitude = abs(value)
    let dominant: Float
    if magnitude.x >= magnitude.y, magnitude.x >= magnitude.z {
        dominant = value.x
    } else if magnitude.y >= magnitude.z {
        dominant = value.y
    } else {
        dominant = value.z
    }
    return dominant < 0 ? -1 : 1
}

private func smoothStep(edge0: Float, edge1: Float, value: Float) -> Float {
    guard edge1 > edge0 else { return value >= edge1 ? 1 : 0 }
    let fraction = clamp((value - edge0) / (edge1 - edge0), 0, 1)
    return fraction * fraction * (3 - 2 * fraction)
}

private extension SIMD3 where Scalar == Float {
    func normalized(or fallback: SIMD3<Float>) -> SIMD3<Float> {
        let magnitude = simd_length(self)
        return magnitude > 0.000_001 ? self / magnitude : fallback
    }
}
